### PR TITLE
feat: Add support for service workers by using indexedDB instead of localStorage

### PR DESCRIPTION
**Pull request by elliotCamblor**
- update several methods to return promises to support await-ing indexedDB connection
-  Add alternative storage strategy for saving to indexedDB rather than window.localStorage. This enables service worker implementations which don't have access to the `window` object. This strategy is invoked when the `self` global is accessible and has the name `ServiceWorkerGlobalScope` 
- consolidate DefaultStorage and DefaultCacheStorage because only one of them was being used anyway
- No longer logging a warning on initialization about not having access to the window object if its been detected that we're running in a service worker environment

How was this tested:
- I ran this against the example app (non-service worker) with both storage strategies succesfully.
- I also installed this local version of the sdk in a test chrome extension successfully (confirmed that indexedDB is being used and anon ID's are being recycled on subsequent initializations. Also configs are being cached)

### CODE REVIEW

File: lib/shared/server-request/package.json [hunk 1]
```diff
@@ -1,5 +0,0 @@
-{
-    "name": "@devcycle/server-request",
```
Comment (jonathannorris) [inline lib/shared/server-request/package.json:2]: why is this removed?

File: sdk/js/src/Client.ts [hunk 2]
```diff
@@ -469,20 +469,20 @@ export class DevCycleClient<
 
             this.onInitialized
```
Comment (jonathannorris) [inline sdk/js/src/Client.ts:470]: it's very weird we use `.then` / `.catch` here instead of `async` / `await`.

File: sdk/js/src/DefaultCacheStore.ts [hunk 3]
```diff
@@ -1,26 +1,39 @@
...
-        this.store?.setItem(storeKey, JSON.stringify(data))
+    constructor(isTesting = false) {
+        super()
+        this.isTesting = isTesting
+        this.init()
+    }
+    async init(): Promise<void> {
+        this.store = this.isTesting ? stubbedLocalStorage : window.localStorage
+    }
+    async save(storeKey: string, data: unknown): Promise<void> {
```
Comment (jonathannorris) [inline sdk/js/src/DefaultCacheStore.ts:26]:
```suggestion
    }
    
    async init(): Promise<void> {
        this.store = this.isTesting ? stubbedLocalStorage : window.localStorage
    }
    
    async save(storeKey: string, data: unknown): Promise<void> {
```

File: sdk/js/src/DefaultCacheStore.ts [hunk 4]
```diff
@@ -1,26 +1,39 @@
...
+    private isTesting: boolean
 
-    save(storeKey: string, data: unknown): void {
-        this.store?.setItem(storeKey, JSON.stringify(data))
+    constructor(isTesting = false) {
+        super()
+        this.isTesting = isTesting
+        this.init()
+    }
+    async init(): Promise<void> {
```
Comment (jonathannorris) [inline sdk/js/src/DefaultCacheStore.ts:23]: is `init()` ever called out-side of the constructor? could this just be in the constructor?

File: sdk/js/src/DefaultCacheStore.ts [hunk 5]
```diff
@@ -32,5 +45,91 @@ const stubbedLocalStorage = {
     key: () => null,
     length: 0,
 }
+// IndexedDB implementation
+export class IndexedDBStrategy extends StorageStrategy {
+    override store: IDBDatabase
+    private ready: Promise<void>
+    private static storeName = 'DevcycleStore'
+    private static DBName = 'DevcycleDB'
+    constructor() {
```
Comment (jonathannorris) [inline sdk/js/src/DefaultCacheStore.ts:54]:
```suggestion
    private static DBName = 'DevcycleDB'
    
    constructor() {
```

Comment (jonathannorris) [inline sdk/js/src/DefaultCacheStore.ts:54]:
```suggestion
    private static storeName = 'DevCycleStore'
    private static DBName = 'DevCycleDB'
    
    constructor() {
```

File: sdk/js/src/DefaultCacheStore.ts [hunk 6]
```diff
@@ -32,5 +45,91 @@ const stubbedLocalStorage = {
...
+
+    // IndexedDB remove
+    async remove(storeKey: string): Promise<void> {
+        await this.ready
+        const tx = this.store.transaction(
+            IndexedDBStrategy.storeName,
+            'readwrite',
+        )
+        const store = tx.objectStore(IndexedDBStrategy.storeName)
+        store.delete(storeKey)
```
Comment (jonathannorris) [inline sdk/js/src/DefaultCacheStore.ts:125]: can you await the delete here? or is it synchronous?

File: sdk/js/src/DefaultCacheStore.ts (hunk 4 above)
Comment (elliotCamblor) [reply sdk/js/src/DefaultCacheStore.ts:23]: its not, I mostly did this because having an `init` function is more readable for the `IndexedDBStrategy` implementation and I wanted both classes to have the same interface. I can change this though

File: sdk/js/src/Client.ts (hunk 2 above)
Comment (elliotCamblor) [reply sdk/js/src/Client.ts:470]: yeah, agreed. I wanted to avoid changing it for the sake of minimizing risk, though I don't love having to mix .then with async/await

File: sdk/js/src/DefaultCacheStore.ts (hunk 6 above)
Comment (elliotCamblor) [reply sdk/js/src/DefaultCacheStore.ts:129]: store.get and store.delete are synchronous. The parent function only needs to be async for the `await this.ready`

**Review by kaushalkapasi — APPROVED** [review]

File: sdk/js/src/Client.ts [hunk 7]
```diff
@@ -235,6 +231,10 @@ export class DevCycleClient<
             }
         }
 
+        this.resolveOnInitialized(this)
```
Comment (ajwootto) [inline sdk/js/src/Client.ts:234]: why move where this is happening? before it only happened if the config fetching process didnt throw

File: sdk/js/src/Client.ts [hunk 8]
```diff
@@ -206,15 +206,11 @@ export class DevCycleClient<
...
-            this.logger.info('Client initialized')
         } catch (err) {
             this.initializeOnConfigFailure(this.user, err)
             return this
         }
 
-        this.eventEmitter.emitInitialized(true)
-
         if (this.user.isAnonymous) {
             this.store.saveAnonUserId(this.user.user_id)
```
Comment (ajwootto) [inline sdk/js/src/Client.ts:215]: if this is async and being ignored it should have `void ` before it

File: sdk/js/src/Client.ts [hunk 9]
```diff
@@ -723,14 +722,13 @@ export class DevCycleClient<
...
                     this.sdkKey,
                     this.logger,
                     this.options,
-                ).then((res) =>
-                    this.logger.info(`Saved response entity! ${res}`),
                 )
+                this.logger.info(`Saved response entity! ${res}`)
             }
 
             this.userSaved = true
```
Comment (ajwootto) [inline sdk/js/src/Client.ts:734]: what are the implications of "this.userSaved" not being synchronously set to true as soon as `setUser` is called, like it was before?

Comment (elliotCamblor) [reply sdk/js/src/Client.ts:734]: great question, ill look into that

File: sdk/js/tsconfig.spec.json [hunk 10]
```diff
@@ -4,11 +4,5 @@
...
         "types": ["jest", "node"]
     },
-    "include": [
-        "**/*.test.ts",
-        "**/*.spec.ts",
-        "**/*.d.ts",
-        "jest.config.ts",
-        "**/__mocks__/*.ts"
-    ]
+    "include": ["**/*.test.ts", "**/*.spec.ts", "**/*.d.ts", "jest.config.ts"]
```
Comment (ajwootto) [inline sdk/js/tsconfig.spec.json:7]: why remove the mocks entry?

File: sdk/js/src/DefaultCacheStore.ts [hunk 11]
```diff
@@ -1,26 +1,41 @@
...
+    constructor(isTesting = false) {
+        super()
+        this.isTesting = isTesting
+        this.init()
     }
 
-    save(storeKey: string, data: unknown): void {
-        this.store?.setItem(storeKey, JSON.stringify(data))
+    async init(): Promise<void> {
+        this.store = this.isTesting ? stubbedLocalStorage : window.localStorage
```
Comment (ajwootto) [inline sdk/js/src/DefaultCacheStore.ts:25]: does this need to be in runtime code? seems like it should be mocked by the tests?

Comment (elliotCamblor) [reply sdk/js/src/DefaultCacheStore.ts:25]: not sure, this is how it was done before

Comment (elliotCamblor) [reply sdk/js/src/DefaultCacheStore.ts:25]: will see if its easy enough to change

File: sdk/js/src/ConfigRequestConsolidator.ts [hunk 12]
```diff
@@ -68,8 +68,8 @@ export class ConfigRequestConsolidator {
                     // don't resolve anything and just make another request while keeping all the previous resolvers
                     this.resolvers.push(...resolvers)
                 } else {
-                    resolvers.forEach(({ resolve }) => resolve(result))
                     this.handleConfigReceivedFunction(result, this.nextUser)
+                    resolvers.forEach(({ resolve }) => resolve(result))
```
Comment (ajwootto) [inline sdk/js/src/ConfigRequestConsolidator.ts:72]: whats the ordering change here doing again?

Comment (elliotCamblor) [reply sdk/js/src/ConfigRequestConsolidator.ts:72]: it was mission critical when I had made the `handleConfigReceivedFunction` async, because it was resolving the promise before all the async operations were complete. Now that were not awaiting the db operations it could go either way. I think it makes more sense to resolve the promise after handling the config being received, but I can also revert it to how it was

Comment (ajwootto) [reply sdk/js/src/ConfigRequestConsolidator.ts:72]: probably true

File: sdk/js/src/Client.ts (hunk 9 above)
Comment (elliotCamblor) [reply sdk/js/src/Client.ts:734]: Looks like its only used to prevent this function from being called multiple times unnecessarily. Dont think theres any real issue here

File: sdk/js/tsconfig.spec.json (hunk 10 above)
Comment (elliotCamblor) [reply sdk/js/tsconfig.spec.json:7]: It was complaining about the fetch mock I think. MIght not be necessary anymore, will double check

Comment (ajwootto) [reply sdk/js/tsconfig.spec.json:7]: removing it would likely give a type error if trying to use `jest` anywhere in a mocks file

File: sdk/js/src/DefaultCacheStore.ts (hunk 11 above)
Comment (elliotCamblor) [reply sdk/js/src/DefaultCacheStore.ts:25]: is it alright if I make this a separate ticket? Since it was already being done this way I'd rather not mess with it in an already risky pr

Comment (elliotCamblor) [reply sdk/js/src/DefaultCacheStore.ts:25]: Made it into a ticket in our triage, Ill try and pick it up during my oncall shift next week

File: sdk/js/src/Client.ts [hunk 13]
```diff
@@ -442,7 +441,7 @@ export class DevCycleClient<
             }
             const config = await this.requestConsolidator.queue(updatedUser)
             if (user.isAnonymous || !user.user_id) {
```
Comment (ajwootto) [inline sdk/js/src/Client.ts:443]: hmm does this logic actually make sense? why is it checking user and then saving updatedUser?

Comment (elliotCamblor) [reply sdk/js/src/Client.ts:443]: I think that's intentional. It's checking whether the user had an id before being updated (which would add an anon user_id if there isnt one already). If it already had a user_id before being updated, then theres no need to save the anon user_id to storage because theres already one in there

Comment (ajwootto) [reply sdk/js/src/Client.ts:443]: I think it's actually checking if the new user (passed in as `user`) is anonymous, and if so it's then saving the anon user id for this new user. But it's using `updatedUser` because the DVCPopulatedUser class is what generates the anon user id if it's not specified.

**Review by ajwootto — APPROVED** [review]
